CMakeLists.txt

```diff
@@ -136,8 +118,45 @@ if(ENABLE_GOLD_LINKER)
     endif()
 endif()
 
-if(CMAKE_BUILD_TYPE MATCHES Release)
+# Explicitly set the build type to Release if no other type is specified
+# on the command line.  Without this, cmake defaults to an unoptimized,
+# non-debug build, which almost nobody wants.
+if(NOT CMAKE_BUILD_TYPE)
+  message(STATUS "No build type specified, defaulting to Release")
+  set(CMAKE_BUILD_TYPE Release)
+endif()
+
+if(CMAKE_BUILD_TYPE MATCHES Debug)
+  message(STATUS "Configuring OSRM in debug mode")
+elseif(CMAKE_BUILD_TYPE MATCHES Release)
   message(STATUS "Configuring OSRM in release mode")
+elseif(CMAKE_BUILD_TYPE MATCHES RelWithDebInfo)
+  message(STATUS "Configuring OSRM in release mode with debug flags")
+elseif(CMAKE_BUILD_TYPE MATCHES MinRelSize)
+  message(STATUS "Configuring OSRM in release mode with minimized size")
+else()
+  message(STATUS "Unrecognized build type - will use cmake defaults")
+endif()
```

Comment:
Users can still do something like `-DCMAKE_BUILD_TYPE=FOO`, which would define CMAKE_BUILD_TYPE, but we don't have any special handling for.  The IF statement is structured to warn when that happens.